cas-server-core-services/src/test/java/org/apereo/cas/services/AbstractResourceBasedServiceRegistryDaoTests.java

```diff
@@ -0,0 +1,391 @@
+
+        final Map<String, String> map = new HashMap<>();
+        map.put("attr1", "newattr1");
+        map.put("attr2", "newattr2");
+        map.put("attr2", "newattr3");
+
+
+        final ReturnMappedAttributeReleasePolicy policy = new ReturnMappedAttributeReleasePolicy();
+        policy.setAllowedAttributes(map);
+        r.setAttributeReleasePolicy(policy);
+
+        final RegisteredService r2 = this.dao.save(r);
+        final RegisteredService r3 = this.dao.findServiceById(r2.getId());
+
+        assertEquals(r, r2);
+        assertEquals(r2, r3);
+        assertNotNull(r3.getAttributeReleasePolicy());
+        assertEquals(r2.getAttributeReleasePolicy(), r3.getAttributeReleasePolicy());
+    }
+
+    @Test
+    public void verifySaveAttributeReleasePolicyAllowedAttrRules() {
+        final RegexRegisteredService r = new RegexRegisteredService();
+        r.setName("testSaveAttributeReleasePolicyAllowedAttrRules");
+        r.setServiceId("testId");
+
+        final ReturnAllowedAttributeReleasePolicy policy = new ReturnAllowedAttributeReleasePolicy();
+        policy.setAllowedAttributes(Arrays.asList("1", "2", "3"));
+        r.setAttributeReleasePolicy(policy);
+
+        final RegisteredService r2 = this.dao.save(r);
+        final RegisteredService r3 = this.dao.findServiceById(r2.getId());
+
+        assertEquals(r, r2);
+        assertEquals(r2, r3);
+        assertNotNull(r3.getAttributeReleasePolicy());
+        assertEquals(r2.getAttributeReleasePolicy(), r3.getAttributeReleasePolicy());
+    }
+
+    @Test
+    public void verifySaveAttributeReleasePolicyAllowedAttrRulesAndFilter() {
+        final RegexRegisteredService r = new RegexRegisteredService();
+        r.setName("testSaveAttributeReleasePolicyAllowedAttrRulesAndFilter");
+        r.setServiceId("testId");
+        r.setTheme("testtheme");
+        r.setEvaluationOrder(1000);
+        r.setAccessStrategy(new DefaultRegisteredServiceAccessStrategy(true, false));
+        r.setProxyPolicy(new RegexMatchingRegisteredServiceProxyPolicy("https://.+"));
+        r.setRequiredHandlers(new HashSet<>(Arrays.asList("h1", "h2")));
+
+        final ReturnAllowedAttributeReleasePolicy policy = new ReturnAllowedAttributeReleasePolicy();
+        policy.setAllowedAttributes(Arrays.asList("1", "2", "3"));
+        r.setAttributeReleasePolicy(policy);
+        r.getAttributeReleasePolicy().setAttributeFilter(new RegisteredServiceRegexAttributeFilter("\\w+"));
+
+        final RegisteredService r2 = this.dao.save(r);
+        final RegisteredService r3 = this.dao.findServiceById(r2.getId());
+
+        assertEquals(r, r2);
+        assertEquals(r2, r3);
+        assertNotNull(r3.getAttributeReleasePolicy());
+        assertEquals(r2.getAttributeReleasePolicy(), r3.getAttributeReleasePolicy());
+    }
+
+    @Test
+    public void verifyServiceType() {
+        final RegexRegisteredService r = new RegexRegisteredService();
+        r.setServiceId("^https://.+");
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [The String literal "^https://.+" appears 8 times in this file; the first occurrence is on line 236](https://www.codacy.com/app/mmoayyed/cas/file/2643672654/issues/source?bid=3256115&fileBranchId=3322154#l236)